### ISSUE TITLE
Add support for read-only accounts

### ISSUE THREAD

**Issue by danfinlay** [MetaMask/metamask-extension#7064]
We should add support for "read only" accounts, for users who want to view their account details, and even browse dapps "from that account's perspective", without carrying any authority over that account.

I thought we had this issue open, but I couldn't find it after [it was requested on twitter by Tim at Truffle](https://twitter.com/timothyjcoulter/status/1165821363484942336).

Should be easy to implement, since [the KeyringController protocol](https://github.com/MetaMask/keyringcontroller) makes it easy to add new key types, and this one has literally no logic to it, it's almost a wonder we don't support it already.

The form for adding one could be similar to private key import, except no private key is required. The keyring would just throw an error for any signing operation, or not implement them (the keyring controller should throw the method-not-found errors).

**Comment by toonsevrin** [comment]
Would be nice indeed.

**Comment by jack0x-tech** [comment]
Yes, would be incredibly helpful for multisig and more advanced, custom Ethereum wallets!

**Comment by toonsevrin** [comment]
Exactly @jack0x-tech, right now we have to write a web3 layer for every single application we want to 'view' from our multisig perspective... I really hope this is given some priority (I don't see any obstacle with implementing this).

**Comment by jack0x-tech** [comment]
@toonsevrin , maybe you guys could consider adding this feature to metamask, then you'd just need to copy/paste the calldata into your multisig "execute" function?

This is our plan if the metamask team can't deliver, but this is maybe 1-2 months off for our team.

Let me know if you guys have any good solutions here, and I'll do the same. Good luck!

**Comment by toonsevrin** [comment]
That sounds perfect. I honnestly have no idea about the metamask source code or how I should go about this from an UI perspective :(

**Comment by lao959** [comment]
Any update on this? @toonsevrin @jack0x-tech 

If no one is working on it, I am planning to implement it. LMK

**Comment by toonsevrin** [comment]
Didn't get to it due to time constraints but would love to see your implementation!

**Comment by lao959** [comment]
sound good. I will try to work on it in the following week. 

Also if anyone do like this feature to be implemented please comment here too, so it will give me more motivation. :-)

**Comment by xJonathanLEI** [comment]
I've got a implementation working locally. Will submit a PR later. However, since new strings will need to be added, I'd love to know how localization is done here. Please kindly advise. Thanks!

**Comment by danfinlay** [comment]
@xJonathanLEI our localization is done using the `t()` function that you see around the codebase. `t('some_string')` will look up a string message as defined in the [locales file for the user's language](https://github.com/MetaMask/metamask-extension/tree/develop/app/_locales). You can start by providing just the `en` locale language, and we'll handle getting it translated.

**Comment by danfinlay** [comment]
A tweet in support of this feature: https://twitter.com/PaulRBerg/status/1420330118577274880

I've opened a community feature request page for this, so people can up-vote it: https://community.metamask.io/t/allow-adding-an-address-as-a-view-only-account/6831

**Comment by fedoracooper** [comment]
Folks, it looks like xJonathanLEI completed an implementation, but he might need help just getting localization working. I found this commit on his fork from late-May:

[Read-only commit](https://github.com/xJonathanLEI/bettermask/commit/3091c7b6149d75a4c6fc52bcfe4a39723f764060)

Is anyone willing to finish up this change so we can get a Watch Wallet working in Metamask?

**Comment by danfinlay** [comment]
I'll bring this up again soon, thanks for the nudge.

**Comment by xJonathanLEI** [comment]
Yes I've got a (quite bare-bone) working implementation that I've been using for debugging (read: stalking) purposes. I'd love to polish it up and submit the PR when I have time (LOL).

**Comment by danfinlay** [comment]
Yes I'm sure time comes readily available to all of us these days ;)

**Comment by ndorf** [comment]
> The keyring would just throw an error for any signing operation, or not implement them (the keyring controller should throw the method-not-found errors).

IMHO it should allow the user to export the unsigned transaction. This would allow the use of dapps with cold wallets.

**Comment by justinjmoses** [comment]
> The keyring would just throw an error for any signing operation, or not implement them (the keyring controller should throw the method-not-found errors).

Qu: instead of this, could the txn be attempted without a valid signature? This could this be utilized in conjunction with a hardhat fork and impersonation to test out dapps while impersonating accounts?

**Comment by artk47** [comment]
@danfinlay is it merged already?

**Comment by Gridzilla** [comment]
Looking forward to this!

**Comment by 0xm3rlin** [comment]
Would appreciate it if this could be added, super helpful to use dApps to compile calls from contracts as well.

**Comment by LaGregance** [comment]
That's a really cool feature and it would be game changer for the metamask mobile app in order to be able to see balances of a ledger address.
Any news ?
